docs/ISD-5865: Charm-architecture (#69)
canonical/mattermost-k8s-operator: docs/reference/charm-architecture.rst

The charm design leverages the `sidecar <https://kubernetes.io/blog/2015/06/the-distributed-system-toolkit-patterns/#example-1-sidecar-containers>`__ pattern to allow multiple containers in each pod with `Pebble <https://documentation.ubuntu.com/pebble/>`__ running as the workload container's entrypoint.

Pebble is a lightweight, API-driven process supervisor that is responsible for configuring processes to run in a container and controlling those processes throughout the workload lifecycle.

Pebble ``services`` are configured through `layers <https://github.com/canonical/pebble#layer-specification>`__, and the following container represents a layer forming the effective Pebble configuration, or ``plan``:

1. A `Mattermost <https://mattermost.com/>`__ container, which runs the Mattermost server application via a startup script that translates charm integration data into Mattermost environment variables.

> As a result, if you run a ``kubectl get pods`` on a namespace named for the Juju model you've deployed the Mattermost charm into, you'll see something like the following:

Reviewer: ```suggestion
As a result, if you run ``kubectl get pods`` on a namespace named for the Juju model you've deployed the Mattermost charm into, you'll see something like the following:
```
nit

.. code:: bash

   NAME                             READY   STATUS    RESTARTS   AGE
   mattermost-k8s-0                 2/2     Running   0          6h4m

This shows there are two containers - the one named above, as well as a container for the charm code itself.

And if you run ``kubectl describe pod mattermost-k8s-0``, all the containers will have as Command ``/charm/bin/pebble``. That's because Pebble is responsible for the processes startup as explained above.

Charm architecture diagram
--------------------------

> Below is a diagram of the application architecture of Mattermost.

Reviewer: Can you add a couple of sentences (it can be brief) that describes what the diagram is showing? If the diagram is visualizing what you've already described in the previous section, then you can reference that text (no need to duplicate)

Author: Done!

.. mermaid::

   C4Container
     System_Boundary(sb_mattermost, "Mattermost Charm") {
       Container_Boundary(cb_app, "App Container") {
         Component(comp_server, "Mattermost Server", "Go Application", "Serves the messaging platform on port 8080")
         Component(comp_start, "start.sh", "Bash Script", "Maps integration data to Mattermost environment variables")
       }
       Container_Boundary(cb_charm, "Charm Container") {
         Component(comp_logic, "Charm Logic", "paas-charm Go Framework", "Controls application deployment & config")
       }
     }

>      Rel(comp_logic, comp_server, "Supervises process via Pebble")
>      Rel(comp_start, comp_server, "Configures & launches")

Reviewer: ```suggestion
     Rel(comp_logic, comp_server, "Supervises process via Pebble")
     Rel(comp_start, comp_server, "Configures & launches")
     UpdateRelStyle(comp_logic, comp_server, $offsetY="10", $offsetX="-60")
     UpdateRelStyle(comp_start, comp_server, $offsetX="20")
```
Nits, to rearrange the labels so that they're easier to read

High-level overview of a Mattermost deployment
----------------------------------------------

> The following diagram shows a typical deployment of the Mattermost charm on a Kubernetes cloud. It consists of the Mattermost charm and a required PostgreSQL charm, with optional integrations for S3 storage, SMTP email, ``OAuth`` authentication, and ingress.

Reviewer: Could you add another sentence that states the consequences of this deployment? You already say that the deployment has the MM charm, required integration to PostgreSQL, and then a bunch of optional integrations. But what's the outcome of all these integrations? It's a Mattermost deployment that's secure (through ingress and authentication), has cloud storage, email delivery, etc.

Author: Done!

.. mermaid::

   C4Container
   title System Architecture: Mattermost Deployment

   Container_Boundary(mattermost_boundary, "Mattermost") {
     System(mattermost, "Mattermost Charm", "Handles team communication and collaboration.")
   }

   System_Boundary(postgres_boundary, "PostgreSQL") {
     SystemDb(postgres, "PostgreSQL", "Required database for storing persistent data.")
   }

   System_Boundary(s3_boundary, "S3 Integrator") {
     SystemDb(s3, "S3 Integrator", "Optional file storage for attachments.")
   }

   System_Boundary(smtp_boundary, "SMTP Integrator") {
     System(smtp, "SMTP Integrator", "Optional email relay service.")
   }

   Person(user, "User", "End user connecting to the Mattermost instance.")

>    Rel(user, mattermost, "connects to")
>    BiRel(mattermost, postgres, "stores data")
>    BiRel(mattermost, s3, "stores files")
>    BiRel(mattermost, smtp, "sends email")

Reviewer: ```suggestion
   Rel(user, mattermost, "connects to")
   BiRel(mattermost, postgres, "stores data")
   BiRel(mattermost, s3, "stores files")
   BiRel(mattermost, smtp, "sends email")
   
   UpdateRelStyle(mattermost, s3, $offsetX="10")
   UpdateRelStyle(mattermost, postgres, $offsetX="-20", $offsetY="10")
```
Nits, to rearrange the labels so they're easier to read

Containers
----------

Configuration files for the container can be found in the respective directory that defines the rock, see the section below.

Mattermost
~~~~~~~~~~

Mattermost is a Go application started via a ``start.sh`` script that maps environment variables provided by the charm integrations (PostgreSQL, S3, SMTP, ``OAuth``) into Mattermost's native ``MM_*`` configuration format.

The Mattermost server listens on port ``8080`` and serves the messaging platform, including its web client, REST API, and WebSocket connections.

The container also includes a set of third-party plugins:

-  |Autolink|_
-  |Matterpoll|_
-  |Giphy|_
-  |Remind|_

.. |Autolink| replace:: :code:`Autolink`
.. _Autolink: https://github.com/mattermost-community/mattermost-plugin-autolink
.. |Matterpoll| replace:: :code:`Matterpoll`
.. _Matterpoll: https://github.com/matterpoll/matterpoll
.. |Giphy| replace:: :code:`Giphy`
.. _Giphy: https://github.com/moussetc/mattermost-plugin-giphy
.. |Remind| replace:: :code:`Remind`

> .. _Remind: https://github.com/scottleedavis/mattermost-plugin-remind

Reviewer: I didn't know about this method to code-format links -- great! 🎉

Author: I didn't know either, thank Claude :D

-  **Loki**: logs can be pushed to the `Loki operator <https://charmhub.io/loki-k8s>`__ via the ``logging`` integration.

Juju events
-----------

For this charm, the following events are observed:

1. ```pebble_ready`` <https://documentation.ubuntu.com/juju/latest/user/reference/hook/#container-pebble-ready>`__: fired on Kubernetes charms when the requested container is ready. Action: check that all required integrations are present and configure the Mattermost container.
2. ```config_changed`` <https://documentation.ubuntu.com/juju/latest/user/reference/hook/#config-changed>`__: usually fired in response to a configuration change using the CLI. Action: validate the configuration and restart the workload.
3. ```update_status`` <https://documentation.ubuntu.com/juju/latest/user/reference/hook/#update-status>`__: periodic event. Action: reconcile the workload state and refresh ingress data.
4. Integration events for ``postgresql``, ``s3``, ``smtp``, and ``oauth``: fired when integration data changes. Action: update the workload configuration and restart the service.

> 5. ```grant_admin_role_action`` <https://charmhub.io/mattermost-k8s/actions>`__: fired when the ``grant-admin-role`` action is executed. Action: Grant the ``system_admin`` role to a user.

Reviewer: these links seem to be broken (at least in the GitHub-rendered file, which doesn't do a great job with reStructuredText files). You may need to play a similar trick as you did above

Author: Fixed it, thanks!

..

   See more in the Juju docs: `Hook <https://documentation.ubuntu.com/juju/latest/user/reference/hook/>`__

Charm code overview
-------------------

The ``src/charm.py`` is the default entry point for a charm and has the ``MattermostK8sCharm`` Python class which inherits from ``paas_charm.go.Charm``.

``paas_charm.go.Charm`` is a base class provided by the `paas-charm <https://github.com/canonical/paas-charm>`__ library, which extends `Ops <https://documentation.ubuntu.com/ops/latest/>`__ (Python framework for developing charms) with built-in support for Go workloads, Pebble service management, and standard integrations (PostgreSQL, S3, ingress, observability).

> The charm itself is minimal, the ``go-framework`` `Charmcraft extension <https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/extensions/>`__ provides the majority of the operational logic, including Pebble layer management, integration handling, and status reporting. Workload-specific configuration is handled by the ``start.sh`` script inside the rock, which converts environment variables set by the charm framework into Mattermost's native ``MM_*`` environment variable format.

Reviewer: ```suggestion
The charm itself is minimal, the ``go-framework`` `Charmcraft extension <https://documentation.ubuntu.com/charmcraft/stable/reference/extensions/>`__ provides the majority of the operational logic, including Pebble layer management, integration handling, and status reporting. Workload-specific configuration is handled by the ``start.sh`` script inside the rock, which converts environment variables set by the charm framework into Mattermost's native ``MM_*`` environment variable format.
```

Pebble ``services`` are configured through `layers <https://github.com/canonical/pebble#layer-specification>`__, and the following container represents a layer forming the effective Pebble configuration, or ``plan``:

1. A `Mattermost <https://mattermost.com/>`__ container, which runs the Mattermost server application via a startup script that translates charm integration data into Mattermost environment variables.

As a result, if you run ``kubectl get pods`` on a namespace named for the Juju model you've deployed the Mattermost charm into, you'll see something like the following:

.. code:: bash

   NAME                             READY   STATUS    RESTARTS   AGE
   mattermost-k8s-0                 2/2     Running   0          6h4m

This shows there are two containers - the one named above, as well as a container for the charm code itself.

And if you run ``kubectl describe pod mattermost-k8s-0``, all the containers will have as Command ``/charm/bin/pebble``. That's because Pebble is responsible for the processes startup as explained above.

Charm architecture diagram

> --------------------------

Reviewer: Nit: I think you can remove this header (no strong opinion)

> Below is a diagram of the application architecture of Mattermost. Mattermost is the main application container, which is configured and managed by the charm logic running in a separate container. The charm logic uses Pebble to interact with the Mattermost process and ensure it is running with the correct configuration, renaming the environment variables to work with Mattermost using `start.sh`.

Reviewer: ```suggestion
Below is a diagram of the application architecture of Mattermost. Mattermost is the main application container, which is configured and managed by the charm logic running in a separate container. The charm logic uses Pebble to interact with the Mattermost process and ensure it is running with the correct configuration, renaming the environment variables to work with Mattermost using ``start.sh``.
```
So that it will be code-formatted (IMO this is the most annoying difference between reStructuredText and Markdown)

.. mermaid::

   C4Container
   title System Architecture: Mattermost Deployment

   Container_Boundary(mattermost_boundary, "Mattermost") {
     System(mattermost, "Mattermost Charm", "Handles team communication and collaboration.")
   }

   System_Boundary(postgres_boundary, "PostgreSQL") {
     SystemDb(postgres, "PostgreSQL", "Required database for storing persistent data.")
   }

   System_Boundary(s3_boundary, "S3 Integrator") {
     SystemDb(s3, "S3 Integrator", "Optional file storage for attachments.")
   }

   System_Boundary(smtp_boundary, "SMTP Integrator") {
     System(smtp, "SMTP Integrator", "Optional email relay service.")
   }

   Person(user, "User", "End user connecting to the Mattermost instance.")

   Rel(user, mattermost, "connects to")
   BiRel(mattermost, postgres, "stores data")
   BiRel(mattermost, s3, "stores files")
   BiRel(mattermost, smtp, "sends email")
   

>    UpdateRelStyle(mattermost, s3, $offsetX="10")

Reviewer: ```suggestion
   UpdateRelStyle(user, mattermost, $offsetY="-20")
   UpdateRelStyle(mattermost, s3, $offsetX="10")
```
Oops sorry one more nit on the label placement

   UpdateRelStyle(mattermost, postgres, $offsetX="-20", $offsetY="10")

Containers
----------

> Configuration files for the container can be found in the respective directory that defines the rock, see the section below.

Reviewer: Nit: You state this information in the section below, so you're safe to remove this line entirely

Author: I was under the impression that every title had to have text beneath them, before another title.

Reviewer: Ahhh yeah that makes sense, Vale used to complain a lot about headers without text underneath. But Vale no longer complains about this, so in the future you shouldn't have to worry so much about it.

But to be fair to the question: Should we have two headers in a row here? Maybe we don't (since we only discuss the Mattermost container). So then the question becomes: Should we discuss the Charm container in more detail? I don't think so, since we discuss the charm code in a separate section.

So, if you really want, you can go ahead and remove the Mattermost header, but I don't have a strong opinion

Author: I united it under one title, and removed the initial line. I agree it is tidier. Thanks

OCI images
----------

We use `Rockcraft <https://canonical-rockcraft.readthedocs-hosted.com/en/latest/>`__ to build the OCI image for Mattermost. The image is defined in the `Mattermost rock <https://github.com/canonical/mattermost-k8s-operator/tree/main/mattermost_rock>`__ and is published to `Charmhub <https://charmhub.io/>`__, the official repository of charms. This is done by publishing a resource to Charmhub as described in the `Charmcraft how-to guides <https://canonical-charmcraft.readthedocs-hosted.com/en/stable/howto/manage-charms/#publish-a-charm-on-charmhub>`__.

Integrations
------------

PostgreSQL
~~~~~~~~~~

PostgreSQL is an open-source object-relational database used by Mattermost to store all persistent data, including users, teams, channels and messages. This is a required integration, the charm will remain in a blocked state until PostgreSQL is integrated.

S3
~~

S3-compatible object storage allows Mattermost to store and retrieve uploaded files (attachments, images, and other media) externally rather than using local filesystem storage. This is an optional integration.

SMTP
~~~~

> SMTP integration enables Mattermost to send outgoing email notifications such as password resets, team invitations, and message alerts through an external SMTP relay. This is an optional integration.

Reviewer: ```suggestion
SMTP enables Mattermost to send outgoing email notifications such as password resets, team invitations, and message alerts through an external SMTP relay. This is an optional integration.
```
or

```suggestion
The SMTP integration enables Mattermost to send outgoing email notifications such as password resets, team invitations, and message alerts through an external SMTP relay. This is an optional integration.
```

``OAuth``
~~~~~~~~~

> ``OAuth`` integration via `Hydra <https://charmhub.io/hydra>`__ enables OpenID Connect-based single sign-on (SSO), allowing users to authenticate through an external identity provider. This is an optional integration.

Reviewer: ```suggestion
The ``OAuth`` integration via `Hydra <https://charmhub.io/hydra>`__ enables OpenID Connect-based single sign-on (SSO), allowing users to authenticate through an external identity provider. This is an optional integration.
```
